[CORE-7095] debug_bundle: Admin API

BenPope opened the pull request:
Add support for:
1. `POST /v1/debug/bundle`
1. `GET /v1/debug/bundle`
1. `DELETE /v1/debug/bundle/{jobid}`
1. `GET /debug/bundle/file/{filename}` (returns `not_implemented`)
1. `DELETE /debug/bundle/file/{filename}` (returns `not_implemented`)

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.2.x
- [ ] v24.1.x
- [ ] v23.3.x

## Release Notes

* none

BenPope commented on src/v/redpanda/admin/api-doc/debug_bundle.def.json:
```diff
@@ -0,0 +1,51 @@
+      "type": "int"
+    },
+    "partition": {
+      "type": "int"
```
Fixing this to a `"string"`

BenPope replied:
Done

BenPope commented on src/v/redpanda/admin/api-doc/debug_bundle.def.json:
```diff
@@ -0,0 +1,51 @@
+  "type": "object",
+  "properties": {
+    "authentication": {
+      "type": "string"
```
Fixing this to:
```json
"scram_authn": {
  "type": "object",
  "required": ["mechanism", "username", "password"],
  "properties": {
    "mechanism": {
      "description": "Scram mechanism",
      "type": "string"
    },
    "username": {
      "description": "username used by RPK to authenticate against Kafka and Admin API",
      "type": "string"
    },
    "password": {
      "description": "password used by RPK to authenticate against Kafka and Admin API",
      "type": "string"
    }
  }
}
```

BenPope replied:
Done

BenPope commented on src/v/redpanda/admin/api-doc/debug_bundle.def.json:
```diff
@@ -0,0 +1,51 @@
+"bundle_start_config_params": {
```
I've added a smattering of descriptions according to the requirements doc.

BenPope replied:
Done

BenPope commented on src/v/redpanda/admin/api-doc/debug_bundle.json:
```diff
@@ -0,0 +1,31 @@
+      "application/json"
+    ],
+    "responses": {
+      "200": {
```
I've added the error codes according to the requirements doc

BenPope replied:
Done

michael-redpanda commented on src/v/redpanda/admin/api-doc/debug_bundle.json:
```diff
@@ -0,0 +1,31 @@
+"/v1/debug/bundle": {
+  "post": {
+    "description": "Posting to this endpoint will trigger the start of an rpk debug bundle process.  If a process is already executing, this endpoint will return a 409 error.  If there is an issue with starting the process, this endpoint will return a 500 error.\n\nNote if job_id is not provided in BundleStartConfig, one will be automatically generated and returned in BundleStartResponse.",
```
The last part (about the Job ID) is incorrect (that's my fault).  The JSON schema indicates that job id is required but the description was wrong.  My bad.  I just updated the API page

BenPope replied:
Done

michael-redpanda replied:
Maybe you updated this later but the description is the old one

michael-redpanda replied:
Ah it got changed in a later commit

michael-redpanda reviewed (commented):
looks great so far

BenPope commented on src/v/debug_bundle/tests/types_test.cc:
```diff
@@ -66,3 +84,26 @@ TEST(types_test, test_partition_print) {
+    EXPECT_FALSE(make("foo/bar/t").has_value());
+    EXPECT_FALSE(make("foo/bar/").has_value());
+    EXPECT_FALSE(make("foo/bar/1d").has_value());
+    EXPECT_EQ(make(" foo/bar/1"), std::nullopt);
+    EXPECT_EQ(make("foo/bar/1 "), std::nullopt);
```
> qq: why the change in pattern for these two?

No idea!

BenPope replied:
Fixed

BenPope commented on src/v/debug_bundle/tests/json_test.cc:
```diff
@@ -0,0 +1,69 @@
+
+    debug_bundle::result<TypeParam> res{outcome::success()};
+
+    EXPECT_NO_THROW(res = from_json<TypeParam>(doc));
```
> question: maybe more appropriate to use `ASSERT_NO_THROW` here since the rest of the test depends on `res`?

Yeah, I already changed a bunch of these.

BenPope replied:
Done

BenPope commented:
Changes in force-push:
1. Added:
  a. `GET /v1/debug/bundle`
  a. `DELETE /v1/debug/bundle/{jobid}`
2. Fixed `status_code`.
3. Improved swagger
4. Addressed review comments on  tests

BenPope commented on src/v/debug_bundle/json.h:
```diff
@@ -11,28 +11,220 @@
+        if (v.IsObject()) {
+            auto o = v.GetObject();
+            scram_creds sc;
+            if (auto r = from_json<decltype(sc.username)>(o, "username", true);
+                r.has_value()) {
+                sc.username = std::move(r).assume_value();
+            } else {
+                return std::move(r).assume_error();
+            }
+            if (auto r = from_json<decltype(sc.password)>(o, "password", true);
+                r.has_value()) {
+                sc.password = std::move(r).assume_value();
+            } else {
+                return std::move(r).assume_error();
+            }
+            if (auto r = from_json<decltype(sc.mechanism)>(
+                  o, "mechanism", true);
+                r.has_value()) {
+                sc.mechanism = std::move(r).assume_value();
```
I think that's up to properly quoting, since `;` is a valid password character.

michael-redpanda replied:
Yeah that's what I'm thinking as well

BenPope commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/754d9c9a988b3a7efb5b63b984a8b853f04cbe6a..99ee73a3a417439f8297f67b27c62c825ee4f57b)
* Added some `operator==`
* Improved test diagnostics
* Improved parsing failure message
* Added test for `debug_bundle_parameters`
* Minor swagger teaks for `"$ref"`
* Added GET and DELETE `/v1/debug/bundle/file/{filename}`

BenPope commented on src/v/debug_bundle/tests/types_test.cc:
```diff
@@ -11,14 +11,32 @@
+struct is_equal {
+    template<typename T>
+    bool operator()(const T& lhs, const T& rhs) const {
+        if constexpr (std::is_same_v<T, partition_selection>) {
```
Done

BenPope commented on src/v/redpanda/admin/server.h:
```diff
@@ -258,6 +268,31 @@ class admin_server {
+              return handler(std::move(req))
+                .then(
+                  [rep = std::move(rep)](json_status_return_type res) mutable {
+                      rep->set_status(res.status);
+                      if (res._body_writer) {
+                          rep->write_body("json", std::move(res._body_writer));
+                      } else {
+                          rep->_content += res._res;
+                      }
+                      return ss::make_ready_future<
+                        std::unique_ptr<ss::http::reply>>(std::move(rep));
```
I couldn't think of a nicer way of wrapping this: https://github.com/redpanda-data/seastar/blob/v24.2.x/include/seastar/http/function_handlers.hh#L99-L111

michael-redpanda replied:
Not questioning that it is necessary, but what makes these endpoints unique to the other ones?

BenPope replied:
> Not questioning that it is necessary, but what makes these endpoints unique to the other ones?

Typically errors are handled like this:
```c++
    if(res.has_error()) {
        ss::httpd::base_exception(
          res.assume_error().message(),
          ss::http::reply::status_type::unprocessable_entity);
    }
```
Which results in:
```json
{"message": "Request body is not a JSON object", "code": 422}
```
So it inextricably links the http status_code to the code in the message.

I'll have a look at implementing another handler type; I'll probably need it to serve a file, anyway.

BenPope replied:
Done

michael-redpanda commented on src/v/redpanda/admin/api-doc/debug_bundle.json:
```diff
@@ -0,0 +1,49 @@
+        "description": "Conflict",
+        "schema": {
+          "$ref": "#/definitions/error_body"
+        }
```
Are we missing a 422 for invalid configuration?

BenPope replied:
> Are we missing a 422 for invalid configuration?
It's not in the API spec for GET, there is no supplied configuration to be invalid, and I can't see anywhere where that error could be produced.

michael-redpanda replied:
Oh weird ok that must have been it, yeah I thought i put the comment on the POST and not the GET.  Computers man

BenPope replied:
> Oh weird ok that must have been it, yeah I thought i put the comment on the POST and not the GET. Computers man

You did. But the diff is greedy, so it "moved".

michael-redpanda commented on src/v/redpanda/admin/api-doc/debug_bundle.json:
```diff
@@ -1,6 +1,6 @@
 "/v1/debug/bundle": {
   "post": {
-    "description": "Posting to this endpoint will trigger the start of an rpk debug bundle process.  If a process is already executing, this endpoint will return a 409 error.  If there is an issue with starting the process, this endpoint will return a 500 error.\n\nNote if job_id is not provided in BundleStartConfig, one will be automatically generated and returned in BundleStartResponse.",
+    "description": "Posting to this endpoint will trigger the start of an rpk debug bundle process.  If a process is already executing, this endpoint will return a 409 error.  If there is an issue with starting the process, this endpoint will return a 500 error.",
```
chore: maybe combine these changes with the previous commit

BenPope replied:
Done

michael-redpanda commented on src/v/redpanda/admin/debug_bundle.cc:
```diff
@@ -0,0 +1,117 @@
+    ss::http::reply::status_type status{ss::http::reply::status_type::ok};
+    switch (ec) {
+    case debug_bundle::error_code::success:
+    case debug_bundle::error_code::debug_bundle_process_running:
```
In the case of `POST /v1/debugbundle/`, process running returns a `409`

michael-redpanda replied:
This is true as well with `/ebug/bundle/file/{file}`

BenPope replied:
Done

Deflaimun commented:
Hi Ben. Sorry for the noob question, but why there's a .def file? 
Should we review docs for both or just the regular .json file?

BenPope commented:
> Hi Ben. Sorry for the noob question, but why there's a .def file? 

Fair question; apart from the error_body, there's not much reuse.

> Should we review docs for both or just the regular .json file? 

Both make sense. Unfortunately the def file isn't valid swagger due to the parsing code.

Please review both. (mostly limited to description, I guess).

kbatuigas commented on src/v/redpanda/admin/api-doc/debug_bundle.def.json:
```diff
@@ -0,0 +1,117 @@
+    }
+  }
+},
+"bundle_start_config": {
```
Sorry @BenPope @michael-redpanda - I saw this [thread](https://github.com/redpanda-data/redpanda/pull/23366#discussion_r1765425721) but wasn't sure if that was implemented or if that might have answered my question. The schema currently indicates that the [request body](https://github.com/redpanda-data/redpanda/pull/23366/files#diff-e2ced4f335d34fdc97bbd85299c4800ae900ae1a3f4a94f8921a47ad5ec9e719R15) must be a `config` object and that object requires a `job_id`. Where does one find the job UUID - is it not the job that will be kicked off by calling the endpoint?

BenPope replied:
> Sorry @BenPope @michael-redpanda - I saw this [thread](https://github.com/redpanda-data/redpanda/pull/23366#discussion_r1765425721) but wasn't sure if that was implemented or if that might have answered my question. The schema currently indicates that the [request body](https://github.com/redpanda-data/redpanda/pull/23366/files#diff-e2ced4f335d34fdc97bbd85299c4800ae900ae1a3f4a94f8921a47ad5ec9e719R15) must be a `config` object and that object requires a `job_id`. Where does one find the job UUID - is it not the job that will be kicked off by calling the endpoint? 

I think the caller must provide a v4 UUID.

kbatuigas replied:
@BenPope do they call a separate endpoint or run an rpk command to retrieve that uuid? (The doc for [`rpk debug`](https://docs.redpanda.com/current/reference/rpk/rpk-debug/rpk-debug/) doesn't have much detail and [`rpk debug bundle`](https://docs.redpanda.com/current/reference/rpk/rpk-debug/rpk-debug-bundle/) doesn't mention a job UUID either.

Maybe related - we only have a handful of the [debug endpoints](https://docs.redpanda.com/api/admin-api/#tag--Debugging) officially documented. I have an [old backlog issue](https://github.com/redpanda-data/documentation-private/issues/1922) to doc debug bundle but I don't know if the PRD and related issues/PRs are out of date. In any case there's a good number of debug endpoints in the [spec](https://github.com/redpanda-data/redpanda/blob/dev/src/v/redpanda/admin/api-doc/debug.json) that we don't have in the docs... Perhaps I should sync with you or @michael-redpanda about it?

michael-redpanda replied:
It's on the caller to generate a valid UUID which will become associated with the created process.

kbatuigas replied:
@michael-redpanda I see, thanks!

kbatuigas commented on src/v/redpanda/admin/api-doc/debug_bundle.json:
```diff
@@ -0,0 +1,250 @@
+},
+"/v1/debug/bundle/{jobid}": {
+  "delete": {
+    "description": "Attempts to kill a running bundle process.  Providing an incorrect job-id returns a 422.  If the process is not currently running will reutrn a 409.  Internal errors return 500",
```
```suggestion
    "description": "Attempts to kill a running bundle process.  Providing an incorrect `job_id` returns a 422.  If the process is not currently running, the request will return a 409 error.  Internal errors will return a 500 status.",
```
Nitpick. Could also go with "job ID".

BenPope replied:
...or one space between a period and the next sentence.

BenPope replied:
>Nitpick. Could also go with "job ID".

I'm going to leave `job_id` since it's a named parameter.

BenPope replied:
Done

BenPope commented on src/v/debug_bundle/types.h:
```diff
@@ -74,6 +74,9 @@ struct partition_selection {
     model::topic_namespace tn;
     absl::btree_set<model::partition_id> partitions;
 
+    static std::optional<partition_selection>
```
Perhaps this should be `result<partition_selection>`, we'd have to include `error.h` from `types.h`.

BenPope commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/99ee73a3a417439f8297f67b27c62c825ee4f57b..8f1db0ea42aad56645e1d9fda8c00af333d28b50)
* Register the handlers using existing mechanism
* Fix error_reply (from `error_code` -> `code`)
* Improve json parsing error reporting
* Fix `debug_bundle_process_running` to set 409 Conflict.
* Improved swagger description

oleiman commented on src/v/debug_bundle/tests/json_test.cc:
```diff
@@ -0,0 +1,257 @@
+using optional_int = std::optional<int>;
+using named_int = named_type<int, struct test_named_tag>;
+
+using JsonTestTypes = ::testing::Types<
+  int,
+  optional_int,
+  named_int,
+  uint64_t,
+  std::chrono::seconds,
+  ss::sstring,
+  uuid_t,
+  special_date,
+  clock::time_point,
+  time_variant,
+  scram_creds,
+  debug_bundle_authn_options,
+  partition_selection,
+  debug_bundle_parameters,
+  std::vector<int>,
+  absl::btree_set<int>>;
+TYPED_TEST_SUITE(JsonTypeTest, JsonTestTypes);
```
this rules

michael-redpanda commented on src/v/redpanda/admin/api-doc/debug_bundle.def.json:
```diff
@@ -73,6 +73,36 @@
+      "description": "Path in API to get the file",
+      "type": "string"
+    },
+    "stdout": {
+      "description": "Only filled in once the process completes.  Content of stdout from rpk",
+      "type": "string"
+    },
+    "stderr": {
+      "description": "Only filled in once the process completes.  Content of stderr from rpk",
+      "type": "string"
+    }
```
stdout and stderr are actually a list of strings

michael-redpanda reviewed (dismissed):
lgtm, we can fix out the stdout/stderr later

BenPope commented on src/v/redpanda/admin/api-doc/debug_bundle.def.json:
```diff
@@ -0,0 +1,117 @@
+      "type": "int"
+    },
+    "logs_since": {
+      "description": "See “rpk debug bundle --help” about the format.  Expecting same format here.",
```
I named the format, with the example from `rpk debug help`.

BenPope commented on src/v/redpanda/admin/api-doc/debug_bundle.def.json:
```diff
@@ -0,0 +1,117 @@
+      "type": "int"
+    },
+    "logs_until": {
+      "description": "See “rpk debug bundle --help” about the format.  Expecting same format here.",
```
I named the format, with the example from `rpk debug help`.

BenPope commented on src/v/redpanda/admin/api-doc/debug_bundle.def.json:
```diff
@@ -0,0 +1,117 @@
+  "required": ["mechanism", "username", "password"],
+  "properties": {
+    "mechanism": {
+      "description": "Scram mechanism",
```
Done

aanthony-rp commented on src/v/redpanda/admin/api-doc/debug_bundle.def.json:
```diff
@@ -0,0 +1,123 @@
+      "description": "When the job was started, in milliseconds since epoch",
+      "type": "int"
+    },
+    "filename": {
```
why do we need a file name?  shouldn't the file just be named job_id.tgz?

BenPope replied:
> why do we need a file name? shouldn't the file just be named job_id.tgz?

Since the UUID is user-supplied, I guess it's possible for the file to already exist, in which case we'd have to create a different filename.

I think this may have been brought up in the RFC.

aanthony-rp replied:
I did read the RFC but I guess I missed this there.  Why do we let the user supply a UUID?  It seems better to automate this, may need to check with Mike on the rationale here.

aanthony-rp commented on src/v/redpanda/admin/api-doc/debug_bundle.json:
```diff
@@ -0,0 +1,256 @@
+    "operationId": "get_debug_bundle_file",
+    "parameters": [
+      {
+        "name": "filename",
```
this API seems to allow fetching any file from any path?   Why expose the file system at all?  we can retrieve the bundle by the bundle ID.

BenPope replied:
> this API seems to allow fetching any file from any path? Why expose the file system at all? we can retrieve the bundle by the bundle ID.

Yes, strong validation would be required at a a minimum here; to ensure that the supplied filename is known to be a debug bundle. That part of the backend has not been written yet.

aanthony-rp reviewed (changes requested):
I have a small concern about the retrieval of bundle files by path/filename.  I think we can design that out to make the API simpler and safer.

BenPope commented on src/v/redpanda/admin/api-doc/debug_bundle.def.json:
```diff
@@ -0,0 +1,117 @@
+      "type": "int"
+    },
+    "logs_since": {
+      "description": "See “rpk debug bundle --help” about the format.  Expecting same format here.",
```
This is now
```
Include logs dated from specified date onward; (journalctl date format: YYYY-MM-DD, 'yesterday', or 'today'). Refer to journalctl documentation for more options.
```
